substrate/frame/staking/src/migrations.rs

```diff
@@ -59,11 +60,42 @@ impl Default for ObsoleteReleases {
 #[storage_alias]
 type StorageVersion<T: Config> = StorageValue<Pallet<T>, ObsoleteReleases, ValueQuery>;
 
+/// Migrating `OffendingValidators` from `Vec<(u32, bool)>` to `Vec<u32>`
+pub mod v15 {
+	use super::*;
+
+	pub struct VersionUncheckedMigrateV14ToV15<T>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for VersionUncheckedMigrateV14ToV15<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let migrated = v14::OffendingValidators::<T>::get()
```

Comment:
You should `::take` from offending validators, or else the old data is never removed.

Reply:
Good catch! Thank you!

Reply:
Fixed. Also added a `post_upgrade` check to ensure the old storage is empty.